### ISSUE TITLE
Enhancement: Hyperparameter experimentation framework

### ISSUE THREAD

**Issue by DrDrewCain** [DrDrewCain/MiniMacLLM#11]
## Problem

Current hyperparameters are hardcoded defaults:
- LoRA rank: 16
- Learning rate: 1e-4
- Batch size: 4
- No systematic way to find optimal values

## Goal

Create framework for hyperparameter experimentation to optimize continual learning performance.

### LoRA Configuration
- **Rank (r):** 4, 8, 16, 32, 64
- **Alpha:** r, 2×r, 4×r
- **Target modules:** All layers vs selective (just attention)
- **Impact:** Parameter efficiency vs learning capacity

### Training Configuration
- **Learning rate:** 1e-5, 3e-5, 1e-4, 3e-4
- **Batch size:** 2, 4, 8, 16 (with gradient accumulation)
- **Warmup steps:** 0, 100, 500
- **Weight decay:** 0.0, 0.01, 0.1

### Continual Learning Specific  
- **Replay ratio:** 0.3, 0.5, 0.7
- **EWC lambda:** 100, 1000, 5000
- **Consolidation frequency:** 500, 1000, 2000 steps

### Generation Parameters
- **Temperature:** 0.5, 0.7, 0.9, 1.0
- **Top-k:** 20, 40, 50
- **Top-p:** 0.8, 0.9, 0.95

### 1. Experiment Configuration
```yaml
# experiments/configs/lora_rank_sweep.yaml
experiment_name: lora_rank_sweep
base_config: configs/medium.yaml

sweep:
  lora_r: [4, 8, 16, 32]
  lora_alpha: [8, 16, 32, 64]

metrics:
  - final_loss
  - perplexity
  - generation_quality
  - training_time
```

### 2. Experiment Runner
```bash
python scripts/run_experiments.py \
  --config experiments/configs/lora_rank_sweep.yaml \
  --output experiments/results/
```

### 3. Results Analysis
```python
# Automatically generate:
- Loss curves comparison
- Best hyperparameter combinations
- Trade-off analysis (quality vs speed vs memory)
- Recommendations for different use cases
```

### Phase 1: Basic Framework
- [ ] Experiment config parser
- [ ] Parameter sweep runner
- [ ] Results logging (CSV/JSON)

### Phase 2: Metrics
- [ ] Track loss, perplexity, time, memory
- [ ] Generation quality scoring
- [ ] Forgetting metrics

### Phase 3: Visualization
- [ ] Plot loss curves
- [ ] Hyperparameter heatmaps
- [ ] Best config recommendations

### Phase 4: Advanced
- [ ] Bayesian optimization (Optuna)
- [ ] Multi-objective optimization
- [ ] Early stopping for bad configs

### Optimal Configs for Different Scenarios

**Memory Constrained:**
```yaml
lora_r: 4
batch_size: 2
gradient_accumulation: 8
```

**Quality Focused:**
```yaml
lora_r: 32
learning_rate: 3e-5
ewc_lambda: 5000
```

**Speed Optimized:**
```yaml
lora_r: 8
batch_size: 16
replay_ratio: 0.3
```

## Acceptance Criteria

- [ ] Experiment framework runs parameter sweeps
- [ ] Results logged with all metrics
- [ ] Visualization tools for analysis
- [ ] Documentation of optimal configs
- [ ] At least 3 experiments completed (LoRA rank, LR, replay ratio)

## Tools to Consider

- **Weights & Biases:** Experiment tracking
- **Optuna:** Hyperparameter optimization
- **Ray Tune:** Distributed experiments
- **Hydra:** Config management

---

**Priority:** Medium - Nice to have after basic quality works  
**Difficulty:** Medium - Framework development effort  
**Impact:** High - Scientific rigor and optimal performance
